django/db/models/query.py

```diff
@@ -570,6 +570,11 @@ def get_or_create(self, defaults=None, **kwargs):
         except self.model.DoesNotExist:
             params = self._extract_model_params(defaults, **kwargs)
             return self._create_object_from_params(kwargs, params)
+        except self.model.MultipleObjectsReturned as e:
+            # This should never be happening in a healthy database, but it does when table index
+            # is corrupt and does not signal existence of a row so a duplicate is added. Provide
+            # an informative message here to help admins save time debugging this rather puzzling situation.
+            raise IntegrityError('get_or_create() returned field not unique, database index may be corrupt: {}'.format(e))
```

Comment:
We have to keep raising `MultipleObjectsReturned` as it changes the function signature and breaks compatiblity. 

This error can also be raised when no unique constraint exists at all on `set(kwargs)` where this error message would be pointing users in the wrong direction. From [the docs](https://docs.djangoproject.com/en/3.0/ref/models/querysets/#get-or-create).

> This method is atomic assuming that the database enforces uniqueness of the keyword arguments (see unique or unique_together). If the fields used in the keyword arguments do not have a uniqueness constraint, concurrent calls to this method may result in multiple rows with the same parameters being inserted.

From my experience a missing unique index is more common than an existing corrupted one. I'm sorry that you had to spend a lot of time figuring out the origin of your issue but that change will likely do more harm than good.

Reply:
@charettes No problem, that's the whole point of review :) Index corruption is indeed rare so I'll just resort to posting an answer to relevant StackOverflow questions.